---
title: "Identification of cancer associated genes using whole genome expression data"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data description

Data was downloaded from the gene expression database (https://www.ebi.ac.uk/gxa/experiments/E-GEOD-19249/Downloads): 

* All the normalized expressions: The original data is a 18307 x 40 matrix. Each row corresponds to a gene. The first three columns are gene ID, gene names and design element, respectively. The rest of columns are assay IDs which correspond to cancer types. 
* Experiment Design data (tsv)

The data contains normalized gene expression levels for 18307 genes for 37 samples. These 37 samples fall into three categories: lung cancer (7 samples), colon cancer (15 samples) and normal (15 samples).

# Questions

1. Which genes are associated with the cancer types?
2. Are there any potential to use specific gene expression data to help diagnose early cancer development?

# Methods & Results

## Import libraries

```{r, warning=FALSE, message=FALSE}
library(plyr) 
library(NMF) 
library(tree) 
library(randomForest) 
library(MASS) 
library(glmnet) 
library(e1071) 
library(class)
```

# Data preprocess

## Data clean

We discarded the gene ID and design element columns. We then converted the assay IDs to cancer types and transposed the data matrix to a 37 x 18307 matrix. And then we added a column of cancer types to the data matrix. The final dataset is a 37 x 18308 matrix with rows corresponding to cancer types and columns corresponding to genes. However, the last column is the cancer types and was used as the response variable in our analysis.

```{r}
# import expressions
expressions = read.csv("data/E-GEOD-19249-A-AFFY-37-normalized-expressions.tsv", sep="\t", stringsAsFactors = FALSE)
# import cancer types
type = read.csv("data/E-GEOD-19249-experiment-design.tsv", sep="\t", stringsAsFactors=F, row.names = 1)[, 2, drop=FALSE]
type = type[colnames(expressions)[-c(1:3)], ]

# transpose expression data so that rows are samples and columns are genes, each gene is feature.
X = t(expressions[, -c(1:3)])
colnames(X) = expressions[, 2] 
rownames(X) = rownames(type)

Y = type

# check predictors and target
dim(X) # 37 rows and 18207 columns
dim(Y) # 37 rows and 1 column
```



# Feature normalization

```{r}
X = scale(X)
```

# Feature selection

We have over 18,000 genes but only 37 samples. We want to select a small set of most important genes for our next analysis. At this step, we implement **F-test** on each gene and select 20 genes that have largest F-value.

```{r}
# F test implementation
# create a data frame with 3 columns to host F-value, p-value and gene name
Fmat = data.frame(Fvalue = rep(NA, ncol(X)),
                  Pvalue = rep(NA, ncol(X)),
                  gene = rep(NA, ncol(X)))

for (i in 1:ncol(X)) {
  Fmat[i, 'gene'] = colnames(X)[i]                   # fill gene name
  aov_fit = aov(X[, i]~Y)                 # run F test
  Fmat[i, 'Fvalue'] = summary(aov_fit)[[1]][1, 'F value']  # fill Fvalue
  Fmat[i, 'Pvalue'] = summary(aov_fit)[[1]][1, 'Pr(>F)']   # fill Pvalue
}
```

We select the top 20 genes with highest Fvalue and pvalue < 0.001

```{r}
library(dplyr)
Fmat = filter(Fmat, Pvalue < 0.001) %>%
  plyr::arrange(desc(Fvalue))
Fmat_top20 = Fmat[1:20, ]
Fmat_top20
```

```{r}
X_top20 = X[, Fmat_top20$gene]
```

```{r}
# linear discriminate analysis
lda_df = data.frame(X_top20, type=as.factor(Y))
lda_fit = lda(type ~ ., data = lda_df)
```


```{r}
library(glmnet)

ridge_fit = glmnet(x = X_top20, y = Y,
                   alpha = 0,
                   standardize = FALSE,
                   family = 'multinomial')

# predict
Y_predict = predict(ridge_fit, newx = X_train_top20, s = log(best_lamb), type='class')
```


